File: clients/ios/StellarChat/StellarChat/ViewModels/ChatViewModel.swift
import Foundation
import SwiftUI

@Observable @MainActor
final class ChatViewModel {
    let groupID: String
    var inputText = ""
    var errorMessage: String?
    var selectedImageData: Data?
    var isSendingImage = false
    var selectedVideoURL: URL?
    var isSendingVideo = false
    var isSendingVoice = false
    /// The message currently being replied to, or nil.
    var replyingToMessage: ChatMessage?
    private weak var appState: AppState?

    var messages: [ChatMessage] {
        let all = appState?.chatMessages[groupID] ?? []
        guard let pinned = appState?.groups.first(where: { $0.id == groupID })?.pinnedEpoch else {
            return all
        }
        return all.filter { $0.epoch == pinned || $0.isSystemMessage }
    }

    var group: ChatGroup? {
        appState?.groups.first(where: { $0.id == groupID })
    }

    var hasBlossomServers: Bool {
        !(appState?.blossomServerURLs.isEmpty ?? true)
    }

    /// ID of the first unread message when the chat was opened, for showing a separator.
    var firstUnreadMessageID: String?

    init(groupID: String, appState: AppState) {
        self.groupID = groupID
        self.appState = appState
        // Capture the first unread message before clearing the count
        let unreadCount = appState.unreadCounts[groupID] ?? 0
        let msgs = appState.chatMessages[groupID] ?? []
        if unreadCount > 0 && msgs.count >= unreadCount {
            firstUnreadMessageID = msgs[msgs.count - unreadCount].id
        }
        // Flush ACKs for messages that arrived while the chat was closed so the sender sees ✓✓.
        appState.sendBacklogAcks(groupID: groupID, count: unreadCount)
        // Mark this group as active and clear unread count
        appState.activeGroupID = groupID
        appState.unreadCounts[groupID] = 0
    }

    func sendMessage() async {
        let text = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }

        let replyToID = replyingToMessage?.id
        do {
            try await appState?.sendMessage(text: text, groupID: groupID, replyToID: replyToID)
            inputText = ""
            replyingToMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func sendImage() async {
        guard let imageData = selectedImageData else { return }
        isSendingImage = true
        defer { isSendingImage = false }

        let replyToID = replyingToMessage?.id
        do {
            try await appState?.sendImage(imageData: imageData, groupID: groupID, replyToID: replyToID)
            selectedImageData = nil
            replyingToMessage = nil
        } catch {
            print("[Blossom] sendImage failed: \(error)")
            errorMessage = error.localizedDescription
        }
    }

    func sendVideo() async {
        guard let videoURL = selectedVideoURL else { return }
        isSendingVideo = true
        defer { isSendingVideo = false }

        let replyToID = replyingToMessage?.id
        do {
            try await appState?.sendVideo(videoURL: videoURL, groupID: groupID, replyToID: replyToID)
            selectedVideoURL = nil
            replyingToMessage = nil
        } catch {
            print("[Blossom] sendVideo failed: \(error)")
            errorMessage = error.localizedDescription
        }
    }

    func sendVoice(audioURL: URL) async {
        isSendingVoice = true
        defer { isSendingVoice = false }

        let replyToID = replyingToMessage?.id
        do {
            try await appState?.sendVoice(audioURL: audioURL, groupID: groupID, replyToID: replyToID)
            replyingToMessage = nil
        } catch {
            print("[Blossom] sendVoice failed: \(error)")
            errorMessage = error.localizedDescription
        }
    }

    func parentMessage(for message: ChatMessage) -> ChatMessage? {
        guard let replyToID = message.replyToID else { return nil }
        return messages.first(where: { $0.id == replyToID })
    }

    func retryMessage(id: String) {
        appState?.retryMessage(groupID: groupID, messageID: id)
    }

    func dismissError() {
        errorMessage = nil
    }

    func onDisappear() {
        if appState?.activeGroupID == groupID {
            appState?.activeGroupID = nil
        }
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/viewmodel/ChatViewModel.kt
package chat.onym.android.viewmodel

import android.content.Context
import android.net.Uri
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import chat.onym.android.model.ChatMessage

class ChatViewModel(
    val groupID: String,
    private val groupListViewModel: GroupListViewModel
) : ViewModel() {
    var inputText by mutableStateOf("")
    var selectedImageUri by mutableStateOf<Uri?>(null)
    var isSendingImage by mutableStateOf(false)
    var selectedVideoUri by mutableStateOf<Uri?>(null)
    var isSendingVideo by mutableStateOf(false)
    var isSendingVoice by mutableStateOf(false)
    var replyingToMessage by mutableStateOf<ChatMessage?>(null)

    val messages: List<ChatMessage>
        get() {
            val allMessages = groupListViewModel.chatMessages[groupID] ?: emptyList()
            val pinned = group?.pinnedEpoch ?: return allMessages
            // When pinned, only show messages from that epoch or system messages
            return allMessages.filter { msg ->
                msg.isSystemMessage || msg.epoch == pinned
            }
        }

    val group: chat.onym.android.model.ChatGroup?
        get() = groupListViewModel.groups.find { it.id == groupID }

    val groupName: String
        get() = group?.name ?: "Chat"

    val inviteLink: String?
        get() {
            val g = group ?: return null
            val code = chat.onym.android.model.InviteCode(
                groupID = g.groupIDData,
                groupSecret = g.groupSecret,
                name = g.name,
                relayHints = g.relayHints,
                members = g.members.toList(),
                epoch = g.epoch,
                salt = g.salt,
                commitment = g.commitment,
                tierRawValue = g.tier.id
            )
            return "https://onym.chat/join?code=${code.encode()}"
        }

    val hasBlossomServers: Boolean
        get() = groupListViewModel.blossomServerURLs.isNotEmpty()

    val isMember: Boolean
        get() {
            val g = group ?: return false
            return groupListViewModel.canSendInGroup(g)
        }

    /** ID of the first unread message when the chat was opened. */
    val firstUnreadMessageID: String?

    init {
        // Capture the first unread message before clearing the count
        val unreadCount = groupListViewModel.unreadCounts[groupID] ?: 0
        val msgs = groupListViewModel.chatMessages[groupID] ?: emptyList()
        firstUnreadMessageID = if (unreadCount > 0 && msgs.size >= unreadCount) {
            msgs[msgs.size - unreadCount].id
        } else null
        // Mark this group as active and clear unread count
        groupListViewModel.activeGroupID = groupID
        groupListViewModel.unreadCounts[groupID] = 0
    }

    fun sendMessage() {
        val text = inputText.trim()
        if (text.isEmpty()) return
        val replyToID = replyingToMessage?.id
        groupListViewModel.sendMessage(groupID, text, replyToID)
        inputText = ""
        replyingToMessage = null
    }

    fun retryMessage(messageID: String) {
        groupListViewModel.retryMessage(groupID, messageID)
    }

    fun sendImage(imageData: ByteArray) {
        isSendingImage = true
        val replyToID = replyingToMessage?.id
        groupListViewModel.sendImage(groupID, imageData, replyToID)
        selectedImageUri = null
        isSendingImage = false
        replyingToMessage = null
    }

    fun sendVideo(context: Context) {
        val uri = selectedVideoUri ?: return
        isSendingVideo = true
        val replyToID = replyingToMessage?.id
        groupListViewModel.sendVideo(groupID, context, uri, replyToID)
        selectedVideoUri = null
        isSendingVideo = false
        replyingToMessage = null
    }

    fun sendVoice(audioFile: java.io.File) {
        isSendingVoice = true
        val replyToID = replyingToMessage?.id
        groupListViewModel.sendVoice(groupID, audioFile, replyToID)
        isSendingVoice = false
        replyingToMessage = null
    }

    fun parentMessage(message: ChatMessage): ChatMessage? {
        val replyToID = message.replyToID ?: return null
        return messages.find { it.id == replyToID }
    }

    override fun onCleared() {
        super.onCleared()
        if (groupListViewModel.activeGroupID == groupID) {
            groupListViewModel.activeGroupID = null
        }
    }
}
